src/Kestrel.Transport.Sockets/Internal/SocketAwaitable.cs

```diff
@@ -52,7 +59,19 @@ public void Complete(int bytesTransferred, SocketError socketError)
         {
             _error = socketError;
             _bytesTransfered = bytesTransferred;
-            Interlocked.Exchange(ref _callback, _callbackCompleted)?.Invoke();
+            var continuation = Interlocked.Exchange(ref _callback, _callbackCompleted);
```

Comment:
So is this basically the equivalent of having a `PipeScheduler.InlineIfAlreadyCompletedOtherwiseGlobalThreadPoolScheduler` if such a thing was possible for the SocketTransport's InputReaderScheduler if such a thing was possible?

If so, shouldn't the [readerScheduler](https://github.com/aspnet/KestrelHttpServer/blob/dev/src/Kestrel.Core/Internal/ConnectionHandler.cs#L89) be switched to PipeScheduler.Inline for the socket transport to avoid double dispatching (or even triple dispatching by the time the data flows through the request body pipe as currently implemented).